rust/lance-io/src/object_store/providers.rs

```diff
@@ -23,70 +27,278 @@ pub trait ObjectStoreProvider: std::fmt::Debug + Sync + Send {
     async fn new_store(&self, base_path: Url, params: &ObjectStoreParams) -> Result<ObjectStore>;
 }
 
+/// A registry of object store providers.
+///
+/// Use [`Self::default()`] to create one with the available default providers.
+/// This includes (depending on features enabled):
+/// - `memory`: An in-memory object store.
+/// - `file`: A local file object store, with optimized code paths.
+/// - `file-object-store`: A local file object store that uses the ObjectStore API,
+///   for all operations. Used for testing with ObjectStore wrappers.
+/// - `s3`: An S3 object store.
+/// - `s3+ddb`: An S3 object store with DynamoDB for metadata.
+/// - `az`: An Azure Blob Storage object store.
+/// - `gs`: A Google Cloud Storage object store.
+///
+/// Use [`Self::empty()`] to create an empty registry, with no providers registered.
+///
+/// The registry also caches object stores that are currently in use. It holds
+/// weak references to the object stores, so they are not held onto. If an object
+/// store is no longer in use, it will be removed from the cache on the next
+/// call to either [`Self::active_stores()`] or [`Self::get_store()`].
 #[derive(Debug)]
 pub struct ObjectStoreRegistry {
-    providers: HashMap<String, Arc<dyn ObjectStoreProvider>>,
-    // TODO: should the registry itself have a cache?
-    // Cache should probable have weak references?
+    providers: RwLock<HashMap<String, Arc<dyn ObjectStoreProvider>>>,
+    // Cache of object stores currently in use. We use a weak reference so the
+    // cache itself doesn't keep them alive if no object store is actually using
+    // it.
+    active_stores: RwLock<HashMap<(String, ObjectStoreParams), Weak<ObjectStore>>>,
```

Comment:
So when will this session be active?  Is it only when we have multiple readers / writers?

Reply:
Yeah this is mainly to address issues when we have a lot of tables open on the same bucket. We might one day cache with TTL, but not sure if it's worth the complexity given that it's not too expensive to reopen the connection.